### DISCUSSION TITLE
LSP syntax highlighting

### DISCUSSION THREAD

**Discussion by kirawi** [helix-editor/helix#5589] [Ideas] [answered]
This will likely require us to refactor the current syntax highlighting system to accommodate other syntax highlighting methods.

**Comment by archseer** [comment]
I'm not a big fan of the LSP highlighting spec. It requires a lot of back and forth traffic with the LSP to do highlighting in real time. VSCode should adapt tree-sitter already..

**Reply by herkhinah** [reply]
tree-sitter isn't fit for all use cases. The tree-sitter approach assumes that there is a static phase distinction between things like type checking and parsing but what about languages where you need type inference to know how to parse the next characters or tokens? What about languages that allow you to extend the syntax of the language with things like parser reflection?
In such use cases you would have to do the work of type checking in both the lsp extension and the tree-sitter implementation.

**Comment by kirawi** [comment]
I plan to make it optional in `config.toml` or `languages.toml`. There are some cases in which tree-sitter isn't enough.

**Comment by pickfire** [comment]
tree-sitter is limited such that it does not have context of the language, like similar returns highlight or similar .await highlights.

**Comment by poliorcetics** [comment]
Also, for Rust, `unsafe` highlighting is nice in VSCode and what I miss the most in helix

**Comment by jakenvac** [comment]
> I'm not a big fan of the LSP highlighting spec. It requires a lot of back and forth traffic with the LSP to do highlighting in real time. VSCode should adapt tree-sitter already..

Does this mean this won't ever be a first class feature of helix? I'm not really fussed either way. I only ask because I'm working on a tool to convert themes between several text editors formats and this would factor into the implementation longer term.

**Comment by LouisGariepy** [comment]
While LSP is heavy, tree-sitter does not provide comparable highlighting. One solution could be to have LSP as an opt-in, with tree-sitter as the default.

As petty as it sounds, the lack of semantic highlighting is one of the very few things that hold me from daily driving Helix. Being able to jump in and see the same theming quality as I'm used to is a must for me, and I imagine other prospective users as well.

**Comment by pickfire** [comment]
@LouisGariepy IIRC in the release notes I remember there is `space-h` select the same symbols in the function, that way it can be similar like part of the syntax highlight feature.

**Comment by LouisGariepy** [comment]
Could a maintainer point out the conditions for an acceptable solution then? If you simply don't want this feature in Helix (which is totally fair!), maybe you could close or mark as "wontfix" to avoid confusion.

To be clear, I'm asking from the perspective of someone who'd like to contribute to this effort. I just don't want to waste my (and your) time with a feature that will never be merged.

**Comment by jakenvac** [comment]
@LouisGariepy 
> Being able to jump in and see the same theming quality as I'm used to is a must for me, and I imagine other prospective users as well.

I'd be interested to see a side by side comparison with the theme/languages you use if you don't mind? In my experience, it's the opposite. I quite often find tree sitter highlighting being better than the alternatives.

**Comment by univerz** [comment]
> I'd be interested to see a side by side comparison with the theme/languages you use if you don't mind?

i miss mutable variables (in rust) with `bold` modifier.

**Comment by archseer** [comment]
That's doable with a custom `locals.scm` query.

**Comment by poliorcetics** [comment]
> That's doable with a custom `locals.scm` query.

Is it though ? They're talking about highlighting the variable as mutable each time it's used, not just at the definition point.

I honestly don't know if this is possible, if it is it seems like a quick win to add it to helix

**Comment by kirawi** [comment]
Yeah, that's doable with a `locals.scm`. It will track the use of an identifier throughout the scope and apply the color it was highlighted with at the definition.

**Comment by LouisGariepy** [comment]
**Edit for future readers:** Please read archseer's response to my comment, which gives a more correct explanation of what is and what is not possible to do with tree-sitter Vs LSP highlighting. My comment contains some inaccuracies, but I'll leave it as is for posterity.

---

@JakeHL 

Alright! I jotted down some basic code to see the potential differences. I tried to emphasize differences due to "limitations", not stylistic ones. There might be more that I'm missing for complex code bases, this is just basic stuff.

Here goes: 

#### Public items are distinct from private ones.
Note that this holds for any item that can be made public, not just functions.

![public_item](https://user-images.githubusercontent.com/68077556/188671692-3c2d0fcb-c1d8-416e-9472-5ead7a514b41.png)

![public_item_helix](https://user-images.githubusercontent.com/68077556/188671698-d9fefdca-d54d-4691-84cf-c8fd674d75d8.png)

#### Mutable variables and operators are distinct from immutable ones.
Mutable functions/method calls are likewise distinct.

![mutable_variables](https://user-images.githubusercontent.com/68077556/188671829-cdaf1baa-e559-48e7-91c5-23a7a3332c27.png)

![mutable_variables_helix](https://user-images.githubusercontent.com/68077556/188671844-129e8cbd-90f5-4491-a8fc-1f7a4386d455.png)

#### Variables in format strings are recognized as such (distinct from the rest of the string).

![format_variables](https://user-images.githubusercontent.com/68077556/188671234-c4ccb5b0-9035-4236-955e-dcacfaf9ddba.png)

![format_variables_helix](https://user-images.githubusercontent.com/68077556/188671305-814cb69e-bc0d-453d-99ab-78ce265aa258.png)

#### Doc comments are distinct from "regular" comments.
Might be hard to see (stylistic choice), but the doc comments are lighter in color.

![doc_comments](https://user-images.githubusercontent.com/68077556/188671903-080d084e-bd14-4627-b380-0a1bb00baa8b.png)

![doc_comments_helix](https://user-images.githubusercontent.com/68077556/188671945-db9cf525-93dd-4153-8b35-4448095da84b.png)

#### Generic types are distinct from their trait bounds.

![trait_bounds](https://user-images.githubusercontent.com/68077556/188671981-5cdfc4a5-d180-46b4-b106-c280285ce926.png)

![trait_bounds_helix](https://user-images.githubusercontent.com/68077556/188672005-a1e6832c-2e3a-4d17-a2d8-38ba864006b7.png)

#### Macro rules binding arguments are distinct from constant patterns.
`$e` and `$es` are distinct in color compared to the rest of the macro_rules syntax.

![macro_rules](https://user-images.githubusercontent.com/68077556/188672195-8f10d94a-11ce-4519-b733-8440ef9fa91e.png)

![macro_rules_helix](https://user-images.githubusercontent.com/68077556/188672204-59b9faf7-acc8-4d31-b7b3-8c933a879c4d.png)

#### Attribute syntax is recognized properly.

![attributes](https://user-images.githubusercontent.com/68077556/188672251-959750a3-0b2b-4407-a781-8d7fddab14a5.png)

![attributes_helix](https://user-images.githubusercontent.com/68077556/188672260-4972d107-1920-4250-8f01-34abeafbbf71.png)


#### Imports (`use`s) are properly highlighted according to their types.

![imports](https://user-images.githubusercontent.com/68077556/188672370-9b223412-04f5-4a76-b40f-2f9723d07c6c.png)

![imports_helix](https://user-images.githubusercontent.com/68077556/188672378-49c688f1-fec3-4bef-9d05-08fa71535f12.png)

#### Trait methods are distinct from non-trait methods.
Notice how the trait method is italicized.

![trait_methods](https://user-images.githubusercontent.com/68077556/188672498-5cedab0d-79ef-4a03-81a3-d2240b429534.png)

![trait_methods_helix](https://user-images.githubusercontent.com/68077556/188672535-e9384728-7231-4172-88cc-0c359222aa51.png)

#### Enum variants are distinct from structs and function calls

![enum_variants](https://user-images.githubusercontent.com/68077556/188673177-1cfb0142-3a6f-48ef-8145-c22a90d4530c.png)

![enum_variants_helix](https://user-images.githubusercontent.com/68077556/188673201-2f4533ca-4b88-4415-917a-216abc630109.png)

# Conclusion
While some of these features might be possible to implement in tree-sitter, I'd like for us to think of the practicality of doing so. I have put many hours into trying to make my "perfect" tree-sitter theme, and I had to give up because I simply could not find a decent/possible way to make some features work. I had a hard time finding documentation and examples of advanced features.

For end users, this results in two things:
1) Being unable to use the plethora of LSP themes that already exist.
2) The themes that do exist in tree-sitter form do not have as extensive highlighting capabilities.

Alternatively, if some of these features are known to be possible in tree-sitter, maybe they should be implemented in one of the default themes, so that users can reuse the same logic without having to reinvent the wheel.

**Comment by kirawi** [comment]
Most of these are currently achievable by modifying the queries except for imports and accurate type highlighting. With https://github.com/helix-editor/helix/issues/1252 however, I think that would be addressed as well.

There are certainly limitations to tree-sitter, such as not being able to highlight markdown in doc comments, but maybe that's something that could be worked on in the future. Language-specific features such as trait methods probably do need LSP highlighting.

**Comment by kirawi** [comment]
It looks like you deleted your last comment, but I think you have a misunderstanding of how tree-sitter works. From how I understand it, tree-sitter a parser generator with an AST geared towards syntax highlighting. However, you also need to write queries to capture AST nodes such that they can be highlighting in themes.

You can check [the tree-sitter playground](https://tree-sitter.github.io/tree-sitter/playground) to see some of what these ASTs look like.

See:
https://tree-sitter.github.io/tree-sitter/syntax-highlighting#queries
https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries

**Comment by jakenvac** [comment]
@LouisGariepy Thanks for that rundown. I think you've demonstrated why it would be a useful feature.

I do think that the main thing it gives us that tree sitter doesn't is type related highlighting. I'm pretty sure with enough time you could create tree-sitter queries for anything that is purely syntax focused.

Here's an example of doc comments & string interpolation already working in Typescript in helix. I might be using a modified Everforest theme, I can't remember.

![Screen Shot 2022-09-07 at 11 10 59](https://user-images.githubusercontent.com/9119973/188853272-2145a5a0-93a4-419c-a64e-0c93bb3b29a0.png)

**Comment by archseer** [comment]
Yeah, tree-sitter grammars are parsers that build an in-memory AST of the file. Then the highlighting queries are like selectors that tag specific parts of the tree with span names, and themes can hook into these spans to highlight them as they wish.

---

> Public items are distinct from private ones.

This is as simple as adding a highlighting query that themes can hook into.

> Mutable variables and operators are distinct from immutable ones.

Can be done with a locals query.

> Variables in format strings are recognized as such (distinct from the rest of the string).

Probably needs a change in the parser, but it could be injected into such strings and the embedded parts interpolated. It works in other languages.

> Doc comments are distinct from "regular" comments.

This is something I've been waiting for https://github.com/tree-sitter/tree-sitter-rust/pull/128

> Generic types are distinct from their trait bounds.

This is as simple as adding a highlighting query that themes can hook into.

> Macro rules binding arguments are distinct from constant patterns.

Not sure yet since macros produce a token tree but probably possible.

> Attribute syntax is recognized properly.

Fixed in https://github.com/helix-editor/helix/commit/301f5d7cf704c2f2e4fc53e258125e39d1845176

> Imports (uses) are properly highlighted according to their types.
> Trait methods are distinct from non-trait methods.

These two are to do so yeah, it could be LSP specific.

> Enum variants are distinct from structs and function calls

Probably possible with locals.

---

I wish LSP provided these highlights as an augmented layer on top of regular highlighting instead of completely overriding highlighting and churning out tons and tons of spans per keypress.

Anyway, for this to be implemented, a vec of produced spans would have to sit somewhere on the document, that would then convert to a `HighlightIter` and hook into our regular rendering mechanism without changes.

**Comment by archseer** [comment]
> I do think that the main thing it gives us that tree sitter doesn't is type related highlighting.

I think in combination with [stack graphs](https://github.blog/2021-12-09-introducing-stack-graphs/) we wouldn't just get go to definition support but also a way to determine the item type since we can peek at the definition

**Comment by kirawi** [comment, marked as answer]
A discussion on how LSP syntax highlighting can be gracefully implemented was discussed on Matrix:

Tree-sitter will be the default highlighter, while range-based LSP syntax highlighting will be merged asynchronously. Additionally, we would compute the correct highlights for locals based on what the LSP syntax highlighter sends.

**Reply by Philipp-M** [reply]
As I was just notified by a comment here, should this be an actual issue again, so that it's obvious, that this could be targeted by a PR?

**Comment by poliorcetics** [comment]
I made https://github.com/helix-editor/helix/pull/6102 for this, if you're interested in the feature don't hesitate to try it out and report bugs :) (or the absence of them, which is always nice)

**Comment by chtenb** [comment]
This is also an important feature for people coming from Visual Studio. For .NET languages, Visual Studio the identifiers of types are highlighted differently depending on their "kind"/"storage". E.g. class, struct, enum, interface. This greatly helps reading code, because the reader instantly sees the storage semantics of the identifiers used in code. Semantic highlighting would enable this set of features in helix 🚀 

FWIW: in Visual Studio the semantic highlighting also lags behind the structural highlighting.

**Comment by Lohann** [comment]
Hi, I came from RustRover and I'm trying to switch to Helix, for me consistent highlithing is very important, follow more examples of tree-sitter limitations:

## Helix
theme: Modified `sonokai`. 

- U256 is a struct (blue), but when imported is wrongly highlighted as a constant (purple).
- Methods when imported are highlighted as a module (white) instead method (green).
<img width="787" alt="Screenshot 2023-10-01 at 12 46 06" src="https://github.com/helix-editor/helix/assets/4323004/150994f0-d479-4b0f-9a04-f4b8e7fb8de5">

## VSCode
theme: default
imports are consistent highlited.
<img width="742" alt="Screenshot 2023-10-01 at 12 52 31" src="https://github.com/helix-editor/helix/assets/4323004/1f9b0744-0043-496d-8a54-39ac4958f6d2">

**Reply by KalinDimitrow** [reply]
I bet this is not the best place to ask my question but I want to know if there is any plans to add support for  semantic highlighting 
It very nice feature especially if you are dyslexic

**Reply by jakenvac** [reply]
@KalinDimitrow For now, I don't see this feature coming to helix. As mentioned [here](https://github.com/helix-editor/helix/pull/6102#issuecomment-1737676445), Archseer has some fairly strong (and justified) opinions on why this shouldn't be implemented in helix.

**Reply by Lohann** [reply]
Update: I abandoned using helix as my main Rust IDE mainly because of this missing feature, but also because it doesn't support copilot, and I can't jumps to inlay-hints types.

**Comment by Quelfth** [comment]
I would like to suggest that the best way to implement this is to add a tree-sitter query predicate like `(#semantic? @capture-name tokenType tokenModifier1 tokenModifier2 ...)` which asks whether a tree-sitter node has particular lsp semantic token applied to it.  I have this working in neovim, and it's really powerful.  It allows you to highlight not only the specific tokens the lsp put the highlight on, but also other related syntax elements.  This allows you to, for instance, give a color each to Functions, Methods, Structs, and Enum Variants, and all of them can have their name match the color of their `( )`.  When you do this for many different kinds of `( )` it becomes rare to find two `))` next to each other which are the same color, which sort of alleviates the need for rainbow brackets, while making the bracket coloration more meaningful.

**Comment by pauladam94** [comment]
It seems like this feature is Not supported by Helix (according to what I have read on this issue). However it is written [here](https://docs.helix-editor.com/languages.html#configuring-language-servers-for-a-language) that it is supported :
<img width="927" height="1346" alt="image" src="https://github.com/user-attachments/assets/0fb5b0af-b5fa-4942-8551-bd65e12d7408" />
Is it an error ? I have not managed to write an lsp that actually gives document highlighting to helix (I am not an expert and this might just be my implementation).

**Reply by maattdd** [reply]
It is *not* marked as supported (the LSP feature is semantic-token, maybe you are confusing it with document-highlight which is very different)

**Comment by SuperMeepEnby** [comment]
Any news on this? If I understand correctly, this is not a "won't do". I'm asking because I wanted to try this editor but it feels hard to read code if I can not distinguish declarations of variables/methods versus their use in the code. 😅
